### PR TITLE
feat: introduce ClusterInfo API [DET-6034]

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

I set out with the intention of having the master pack a bunch of information into the container as a file, rather than as environment variables.  However, that was a lot of churn for no real benefit, so instead I expanded the `fetch_context` step in `entrypoint.sh` to just save a bunch of environment variables to file, so they can easily be re-read later without issue.

A significant side-effect of this PR is that the environment variables read by the ClusterInfo will effectively become part of the master-harness APIs, which we want to eventually version.

## Commentary (optional)

The astute reviewer may realize that the ClusterInfo object looks suspiciously similar to the old EnvContext object.  In fact, they are very similar in their construction and similar in purpose.  When writing a launch layer, it is natural to need a convenient way to find out information about how the cluster started your task.  However, we made a bad mistake in passing the EnvContext into the lowest layers of the harness.  The results were weak abstractions in components of the harness, difficult mocking required for testing, and the inability to meaningfully deliver a convenient local training mode.

So the ClusterInfo is basically what EnvContext should have been; a clean API for finding out information about the environment you are running in.  But we need to warn users to _never_ pass the ClusterInfo object from their launch layer to their training script, or they will have trouble ever running their code outside of the Determined cluster.

### CODE REVIEW

File: harness/determined/common/storage/azure.py [hunk 1]
```diff
@@ -30,73 +31,69 @@ def __init__(
...
+            # Use posixpath so that we always use forward slashes, even on Windows.
+            container_blob = posixpath.join(self.container, storage_prefix, rel_path)
+            blob_dir, blob_base = posixpath.split(container_blob)
             abs_path = os.path.join(storage_dir, rel_path)
+            logging.debug(f"Uploading blob {blob_base} to container {blob_dir}.")
+            self.client.put(blob_dir, blob_base, abs_path)
 
-            os.makedirs(os.path.dirname(abs_path), exist_ok=True)
+    @util.preserve_random_state
+    def download(self, uuid: str, storage_dir: str) -> None:
```
Comment (ioga) [inline harness/determined/common/storage/azure.py:71]: how about calling `uuid` args throughout these functions `checkpoint_uuid`/`ckpt_uuid` instead?

I've experienced APIs where there're various types of objects both being identified with `uuid`s. e.g. hypothetical checkpoints and tasks where a task makes a checkpoint. all typed as strings. It's confusing to deduce which uuid you need to pass to some functions. having explicit name like `task_uuid` and `checkpoint_uuid` helps.

after all, the second argument to this method is `storage_dir`, not `dir`.

Comment (rb-determined-ai) [reply harness/determined/common/storage/azure.py:71]: Ok.  I actually changed these from being called `storage_id` everywhere to `uuid` everywhere, but your reasoning makes sense; I will switch back to `storage_id`.  (I'll do it in the `uuids` pr though, https://github.com/determined-ai/determined/pull/2970).

File: harness/determined/_info.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+        # _config has no getter because it's not a stable API; as the experiment config version
+        # increases, the shape of the experiment config dict might change as the master is upgraded.
+        # As a result, we should strongly discourage users from reading the experiment config dict
+        # in their training code.  They should focus on the data field and the hparams field, which
+        # have always been and will always be user-defined.
+        self._config = config
+
+    @property
+    def trial_id(self) -> int:
+        return self._trial_id
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_info.py:76]: I have increasingly seen external libraries that use the `@property` decorator in all cases, so I'm not sure that it's exactly _un_-pythonic.

That said, I'm not sold on it myself.  The cons I see are:
- I don't like the extra verbosity.
- It takes something simple and makes it not simple

But the pros I see are:
- I do like the read-only nature of properties; it shows our intentions for the API very clearly, if any users ever casually try to set it or if they read the code.  Users who read the source code to find `._trial_id` and set that will eventually get what they had coming when their code breaks... there's no preventing that really.
- I like that us having a function getter is more flexible in the future, if we e.g. want to rename a property, and leave the old symbol present but with a deprecation warning.  And in that potential future case, it's nice that the property was read-only since the beginning.

I think the pros outweigh the cons.  I especially like the read-only-ness because in compiled languages I like `const` whenever possible.  But I don't think I care all that much, so if you feel strongly I can be ok with normal attributes rather than `@property` attributes.  Just let me know.

File: harness/determined/common/api/authentication.py [hunk 3]
```diff
@@ -10,18 +10,20 @@
...
 Credentials = NamedTuple("Credentials", [("username", str), ("password", str)])
 
 PASSWORD_SALT = "GubPEmmotfiK9TMD6Zdw"
 
-_cur_allocation_token = os.environ.get("DET_ALLOCATION_SESSION_TOKEN", "")
-
 
 def get_allocation_token() -> str:
-    return _cur_allocation_token
+    info = det.get_cluster_info()
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/common/api/authentication.py:23]: Yes, that seems good to me.

File: harness/determined/_execution.py [hunk 4]
```diff
@@ -21,12 +21,12 @@ def __init__(self, msg: str = "...") -> None:
 
 
 def _get_gpus(limit_gpus: Optional[int]) -> Tuple[bool, List[str], List[int]]:
-    gpu_ids, gpu_uuids = gpu.get_gpu_ids_and_uuids()
+    gpu_uuids = gpu.get_gpu_uuids()
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_execution.py:24]: sgtm

File: harness/determined/_info.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ioga) [reply harness/determined/_info.py:76]: >I like that us having a function getter is more flexible in the future, if we e.g. want to rename a property, and leave the old symbol present but with a deprecation warning. And in that potential future case, it's nice that the property was read-only since the beginning.

if you have a plain simple field `self.trial_id = trial`, and you want to deprecate it, you can add the `@property` getter for it specifically with the deprecation warning at that time. making them `@properties` ahead of time doesn't add to futureproofing.

>Users who read the source code to find ._trial_id and set that will eventually get what they had coming when their code breaks... there's no preventing that really.

As a source code reader for 3p library code, I'd prefer cleaner and concise code, and I won't make assumptions it's safe to write into it unless the documentation says so.

Read-only-ness doesn't cut it for me. I'd vote against properties.

File: harness/determined/_info.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    def gpu_uuids(self) -> List[str]:
+        return self._gpu_uuids
+
+    @classmethod
+    def _by_inspection(cls) -> "ResourcesInfo":
+        return cls(gpu_uuids=gpu.get_gpu_uuids())
+
+    def _to_file(self, path: str = "/run/determined/info/resources.json") -> None:
+        with open(path, "w") as f:
+            json.dump({k.lstrip("_"): v for k, v in vars(self).items()}, f)
```
Comment (azhou-determined) [inline harness/determined/_info.py:128]: would it be helpful to have a to_json classmethod to serialize?

File: harness/determined/_info.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+
+    @property
+    def gpu_uuids(self) -> List[str]:
+        return self._gpu_uuids
+
+    @classmethod
+    def _by_inspection(cls) -> "ResourcesInfo":
+        return cls(gpu_uuids=gpu.get_gpu_uuids())
+
+    def _to_file(self, path: str = "/run/determined/info/resources.json") -> None:
```
Comment (azhou-determined) [inline harness/determined/_info.py:126]: do we need a check if file exists?

File: harness/determined/_info.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+            unique_port_offset=int(os.environ["DET_UNIQUE_PORT_OFFSET"]),
+            inter_node_network_interface=os.environ.get("DET_INTER_NODE_NETWORK_INTERFACE"),
+        )
+
+    def _to_file(self, path: str = "/run/determined/info/trial.json") -> None:
+        with open(path, "w") as f:
+            json.dump({k.lstrip("_"): v for k, v in vars(self).items()}, f)
+
+    @classmethod
+    def _from_file(cls, path: str = "/run/determined/info/trial.json") -> Optional["TrialInfo"]:
```
Comment (azhou-determined) [inline harness/determined/_info.py:107]: what do you think about using pathlib/Path type for these? if nothing else, makes checking .exists() really straightforward

File: harness/determined/_info.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+        self._unique_port_offset = unique_port_offset
+        # TODO: Get rid of this in favor of launch layer configs?
+        self._inter_node_network_interface = inter_node_network_interface
+
+    @classmethod
+    def _from_env(cls) -> "TrialInfo":
+        assert "DET_TRIAL_ID" in os.environ, "must be run inside a Trial container"
+        experiment_config = json.loads(os.environ["DET_EXPERIMENT_CONFIG"])
+        return cls(
+            trial_id=int(os.environ["DET_TRIAL_ID"]),
```
Comment (azhou-determined) [inline harness/determined/_info.py:90]: what if we had a class for all these DET environment variables? could read by defined key consts instead of arbitrary strings and also enforce consistent read types like json.loads or ints

File: harness/determined/exec/prep_container.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import determined as det
+from determined import constants
+from determined.common.api import certs, request
+
+
+def trial_prep(info: det.ClusterInfo, cert: certs.Cert) -> None:
+    trial_info = det.TrialInfo._from_env()
+    trial_info._to_file()
+
+    resp = request.get(info.master_url, f"api/v1/experiments/{trial_info.experiment_id}/model_def")
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/exec/prep_container.py:16]: because the server might provide a cert.  This is the case with _all_ tls masters, though technically its only necessary in self-signed certificate situations.

File: harness/determined/gpu.py [hunk 10]
```diff
@@ -70,23 +68,10 @@ def get_gpus() -> List[GPU]:
...
-        _, gpu_uuids = get_gpu_ids_and_uuids()
-        if not gpu_uuids:
-            raise det.errors.InternalException("Failed to find GPUs for GPU-only trial")
 
-        if slot_ids is not None:
-            check.equal_lengths(slot_ids, gpu_uuids, "Mismatched slot_ids and container_gpus.")
-        return gpu_uuids
-    return []
+def get_gpu_uuids() -> List[str]:
+    return [gpu.uuid for gpu in sorted(get_gpus(), key=lambda gpu: gpu.id)]
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/gpu.py:77]: gpus have a "natural" OS-defined order, so a uuid list that isn't ordered that way would be very obnoxious.  That ordering is the `gpu.id` value.

File: harness/determined/_info.py (hunk 6 above)
Comment (rb-determined-ai) [reply harness/determined/_info.py:126]: I would say, "no".  `Thing._to_file(path)` should write to `path` unconditionally.

File: harness/determined/_info.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/determined/_info.py:128]: I can add a helper function, but it will be a standalone function.  I think that inheritance hierarchies which only exist for code sharing purposes result in messy object-oriented code:  E.g. a TrialInfo is not "like" a ClusterInfo object.

File: harness/determined/exec/prep_container.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    resp.raise_for_status()
+
+    tgz = base64.b64decode(resp.json()["b64Tgz"])
+
+    with tarfile.open(fileobj=io.BytesIO(tgz), mode="r:gz") as model_def:
+        # Ensure all members of the tarball resolve to subdirectories.
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/exec/prep_container.py:22]: this is a security concern; you don't want somebody giving you a tarball with a path that's going to try to overwrite `../../../../../../../../etc/sudoers` or something like that.

As I recall, this was a check that the `tarfile` helper function normally applied, so I added it here since I couldn't use the helper function for some reason or another.  Right now this part of the code just moved from `fetch_context.py`.

File: harness/determined/_info.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+            return cls(
+                trial_info=TrialInfo._from_file(),
+                rendezvous_info=RendezvousInfo._from_file(),
+                resources_info=ResourcesInfo._from_file(),
+                **json.load(f),
+            )
+
+    @property
+    def latest_checkpoint(self) -> Optional[str]:
+        if self.task_type != "TRIAL":
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_info.py:267]: Mixed.  On the one hand, it's probably perfectly fine.

On the other hand, "Enum" implies that there's a fixed set that you can reasonably expect to never see values outside of.  But eventually we want to version the master-harness APIs such that the master can be updated and old user images (aka old harness versions) still run.  And in that world, it's perfectly reasonable that the `info.task_type` might contain unknown values, written by a master that is a version ahead of the harness code that is executing.

I think there's a fancy explanation for this: https://stackoverflow.com/a/4605019

A more accurate pattern would be for task_type to be a class, and the trial_task_type to be a subclass of task_type or something.  Because you can always have new subclasses; it's not normal for code to try to enumerate all possible subclasses of PyTorchTrial.  But that is the wrong pattern for other reasons...

File: harness/determined/_info.py (hunk 8 above)
Comment (rb-determined-ai) [reply harness/determined/_info.py:90]: I'm a little confused... I think `ClusterInfo` _is_ the class for all of these environment variables.  What did you have in mind?

File: harness/determined/_info.py (hunk 7 above)
Comment (rb-determined-ai) [reply harness/determined/_info.py:107]: > makes checking .exists() really straightforward

`os.path.exists()` is equally straight-forward.  Personally I think `pathlib.Path` is silly, but if you feel strongly I can use it here.  It's not user facing so it doesn't really matter.

File: harness/determined/_train_context.py [hunk 13]
```diff
@@ -277,22 +279,30 @@ class DistributedContext:
...
+        if self._info.cross_size > 1:
+            if chief_ip is None:
+                raise AssertionError(
+                    f"rank_info has cross_size ({self._info.cross_size}) but chief_ip was not "
+                    "provided.  When cross_size > 1, the chief_ip parameter is required."
+                )
+            self._chief_ip = chief_ip
+        else:
+            # When cross_size == 1, always contact the chief as localhost.
+            self._cheif_ip = "127.0.0.1"
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_train_context.py:305]: damn that word.  Still gets me sometimes.

File: harness/determined/_info.py (hunk 8 above)
Comment (azhou-determined) [reply harness/determined/_info.py:100]: I was picturing some kind of master class that contained all the environment variables we used and some getters (maybe setters) that would give consistent return types. if the environment variables should only be used in context of ClusterInfo and RendezvousInfo and TrialInfo and not standalone, then it's pointless. but like if i want just DET_TRIAL_ID in some other part of the harness, maybe it would be nice if there was a single API for getting it? like env.trial_id or something. basically never having to pass in an arbitrary environment variable string and casting it to a type if used outside of the ...Info objects.

Comment (rb-determined-ai) [reply harness/determined/_info.py:100]: > like env.trial_id or something

lol like `info.trial.trial_id`?  That's like the stated purpose of this class.

But in general, you shouldn't be reading these values at random points in the harness; that's how you build fragile code that only runs inside of our cluster.  Which is exactly what we did, which is exactly why we still haven't shipped local training mode.  _Everything_ assumes it can access values that are only set when on the cluster.  There's no clean interface anywhere.

Comment (azhou-determined) [reply harness/determined/_info.py:100]: but the current structure means to read a singular value you'd have to read a batch of them related to trialInfo or clusterInfo. but your explanation of why they shouldn't be read this way makes sense.

File: harness/determined/_info.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/_info.py:275]: i'd never thought of that. good read, thanks.

**Review by azhou-determined — APPROVED** [review]

File: harness/determined/_execution.py [hunk 14]
```diff
@@ -21,12 +21,14 @@ def __init__(self, msg: str = "...") -> None:
...
+    gpus = gpu.get_gpus()
+
     if limit_gpus is not None:
-        use_gpu = len(gpu_uuids) > 0 and limit_gpus > 0
-        return use_gpu, gpu_uuids[:limit_gpus], gpu_ids[:limit_gpus]
-    use_gpu = len(gpu_uuids) > 0
-    return use_gpu, gpu_uuids, gpu_ids
+        gpus = gpus[:limit_gpus]
+
+    use_gpus = len(gpus) > 0
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_execution.py:29]: use_gpus can be forced to be false by the caller setting `limit_gpus=0`.

It's a pretty clumsy interface to configure the training code, but that's just how it is with `EnvContext`.  Refactoring EstimatorTrial and TFKerasTrial which consume `EnvContext` is not worth the cost right now (maybe not ever), so having some awkwardness here to keep the legacy code working is acceptable, IMO.

File: harness/determined/_info.py [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+        self.container_rank = container_rank
+
+    def _to_file(self, path: str = DEFAULT_RENDEZVOUS_INFO_PATH) -> None:
+        """
+        to_file writes the RendezvousInfo to a well-known location in a Determined container.  This
+        is called automatically early in the lifetime of a container, so user code can always expect
+        this file to be written before user code runs.
+        """
+        with open(path, "w") as f:
+            json.dump(vars(self), f)
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_info.py:42]: It doesn't matter, they would have the same effect.

If `as_dict` did something special or tricky, we could talk about putting it somewhere that it could be accessed from other code, making sure it was uniformly used across the codebase, writing tests for it, etc.  But with how `as_dict` is so simple I'm not even sure having that having it outweighs not having it (DRY principle vs decreasing layers of indirection).

File: harness/determined/_info.py [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+
+    @property
+    def latest_checkpoint(self) -> Optional[str]:
+        if self.task_type != "TRIAL":
+            raise RuntimeError(
+                "the .latest_checkpoint property is not yet supported when .task_type "
+                f'("{self.task_type}") != "TRIAL"'
+            )
+        if self._latest_checkpoint is None:
+            return None
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/_info.py:281]: good catch

**Comment by rb-determined-ai** [comment]
upstream tests passed (https://github.com/determined-ai/determined/pull/3026)
